cli/flox/src/commands/mod.rs

```diff
@@ -329,14 +343,210 @@ impl FloxArgs {
-            Commands::Internal(group) => group.handle(config, flox).await?,
+        let result = match self.command.unwrap() {
+            Commands::Help(group) => {
+                group.handle();
+                Ok(())
+            },
+            Commands::Development(group) => group.handle(config, flox).await,
+            Commands::Sharing(group) => group.handle(config, flox).await,
+            Commands::Additional(group) => group.handle(config, flox).await,
+            Commands::Internal(group) => group.handle(config, flox).await,
+        };
+
+        // This will print the update notification after output from a successful
+        // command but before an error is printed for an unsuccessful command.
+        // That's a bit weird,
+        // but I'm not sure it's worth a refactor.
+        match check_for_update_handle.await {
+            Ok(update_notification) => {
+                UpdateNotification::handle_update_result(update_notification);
+            },
+            Err(e) => debug!("Failed to check for CLI update: {}", display_chain(&e)),
+        }
+
+        result
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+struct LastUpdateNotification {
+    #[serde(with = "time::serde::iso8601")]
+    last_notification: OffsetDateTime,
+}
+
+#[derive(Debug, PartialEq)]
+struct UpdateNotification {
+    new_version: String,
+    notification_file: PathBuf,
+}
+
+#[derive(Debug, Error)]
+enum UpdateNotificationError {
+    /// If someone can't check for updates because of a network error, we'll
+    /// want to silently ignore it.
+    #[error("network error")]
+    Network(#[source] reqwest::Error),
+    /// If someone can't check for updates because of an IO error, we'll want to
+    /// silently ignore it.
+    #[error("IO error")]
+    Io(#[source] io::Error),
+    /// Other errors indicate something we didn't expect may have happened,
+    /// so we want to report it with Sentry.
+    #[error(transparent)]
+    WeMayHaveMessedUp(#[from] anyhow::Error),
+}
+
+impl UpdateNotification {
+    pub async fn check_for_and_print_update_notification(cache_dir: impl AsRef<Path>) {
+        Self::handle_update_result(Self::check_for_update(cache_dir).await)
+    }
+
+    /// If the user hasn't been notified of an update after
+    /// UPDATE_NOTIFICATION_EXPIRY time has passed, check for an update.
+    pub async fn check_for_update(
+        cache_dir: impl AsRef<Path>,
+    ) -> Result<Option<Self>, UpdateNotificationError> {
+        let notification_file = cache_dir.as_ref().join(UPDATE_NOTIFICATION_FILE_NAME);
+        Self::testable_check_for_update(
+            notification_file,
+            Self::get_latest_version,
+            UPDATE_NOTIFICATION_EXPIRY,
+        )
+        .await
+    }
+
+    /// If the user hasn't been notified of an update after `expiry` time has
+    /// passed, check for an update.
+    async fn testable_check_for_update<Fut>(
```

Comment:
I'll look at passing a Future instead.

For `testable_check_for_update` and `check_for_and_print_update_notification` it basically sounds like you don't like small wrapper functions that make the pub API simpler. And you don't mind boilerplate/ (joining a path, passing a constant in multiple places). I'd rather trim the boilerplate and make the pub API simpler. Is there any actual downside to that other than personal preference/smells bad?

Reply:
Passed the Future - that's definitely cleaner. TIL `async { value }` is a thing

Reply:
> TIL `async { value }` is a thing

Just FYI you could also do `futures::ready(value)` (not recommending it here, just spreading knowledge).

> it basically sounds like you don't like small wrapper functions that make the pub API simpler.

Not exactly. I prefer functions that serve one purpose (to the degree that it's reasonable, I wouldn't break up `ensure_registered` into individual functions).

> Is there any actual downside to that other than personal preference/smells bad?

I think my knee-jerk reaction was to the name `testable_*` because that implies the other function _isn't_ testable and an untestable API is a bad sign. Let's just call it `check_for_update_inner` and move on.

Reply:
> Not exactly. I prefer functions that serve one purpose (to the degree that it's reasonable, I wouldn't break up `ensure_registered` into individual functions) i.e. if it has "and" in the name it's a smell.

Makes sense

> > Is there any actual downside to that other than personal preference/smells bad?
> 
> I think my knee-jerk reaction was to the name `testable_*` because that implies the other function _isn't_ testable and an untestable API is a bad sign. Let's just call it `check_for_update_inner` and move on.

👍 done